### PR TITLE
add custom collective transforms

### PR DESCRIPTION
**Pull request by CosmoMatt**
@jasonmcewen this PR builds on the Fourier-Wigner precompute transform work and implements custom transforms which collectively transform: 

- [x] Subsets of Wigner coefficients (collection of spin spherical harmonic coefficients) to collection of spin signals on sphere
- [x] Collection of spin signals on sphere to their spin spherical harmonic coefficients (stable wrt spin and can accept duplicated spins as per multimodal data)
- [x] Signal on SO(3) on a subset of its Wigner coefficients (subset of spin spherical harmonic coefficients).

These are all the core transforms we will need going forward.

### CODE REVIEW

File: s2fft/precompute_transforms/custom_ops.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+    Returns:
+        np.ndarray: A collection of spin signals with shape :math:`[batch, n_s, n_\theta, n_\phi, channels]`.
+
+    """
+    if sampling.lower() not in ["mw", "mwss"]:
+        raise ValueError(
+            f"Fourier-Wigner algorithm does not support {sampling} sampling."
+        )
+
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:44]: We should check the input shapes are as expected, e.g. that flmn and spins have the same n_s and that shapes match the L provided.

File: s2fft/precompute_transforms/custom_ops.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    spins: np.ndarray,
+    DW: Tuple[np.ndarray, np.ndarray],
+    L: int,
+    sampling: str = "mw",
+) -> np.ndarray:
+    r"""
+    Transforms an arbitrary subset of Wigner coefficients onto a subset of spin signals
+    on the sphere.
+
+        This function takes a collection of spin spherical harmonic coefficients each with
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:20]: Indented too much.

File: s2fft/precompute_transforms/custom_ops.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+            with shape :math:`[batch, n_s, L, 2L-1, channels]`.
+        spins (np.ndarray): Spins of each field in fs with shape :math:`[n_s]`.
+        DW (Tuple[np.ndarray, np.ndarray]): Fourier coefficients of the reduced
+            Wigner d-functions and the corresponding upsampled quadrature weights.
+        L (int): Harmonic band-limit.
+        sampling (str, optional): Sampling scheme.  Supported sampling schemes include
+            {"mw", "mwss"}. Defaults to "mw".
+
+    Returns:
+        np.ndarray: A collection of spin signals with shape :math:`[batch, n_s, n_\theta, n_\phi, channels]`.
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:37]: Line too long.

File: s2fft/precompute_transforms/custom_ops.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    L: int,
+    sampling: str = "mw",
+) -> np.ndarray:
+    r"""
+    Transforms an arbitrary subset of Wigner coefficients onto a subset of spin signals
+    on the sphere.
+
+        This function takes a collection of spin spherical harmonic coefficients each with
+        a different (though not necessarily unique) spin and maps them back to their
+        corresponding pixel-space representations. Following this operation one may
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:22]: I'm not sure it is correct to say "exploiting the correct Mackey functions".  Perhaps just delete this sentence, here and elsewhere, since this is related to downstream uses.

File: s2fft/precompute_transforms/custom_ops.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+    spins: jnp.ndarray,
+    DW: Tuple[jnp.ndarray, jnp.ndarray],
+    L: int,
+    sampling: str = "mw",
+) -> jnp.ndarray:
+    r"""
+    Transforms an arbitrary subset of Wigner coefficients onto a subset of spin signals
+    on the sphere (JAX).
+
+        This function takes a collection of spin spherical harmonic coefficients each with
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:94]: Incorrect indentation.

File: s2fft/precompute_transforms/custom_ops.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+        a different (though not necessarily unique) spin and maps them back to their
+        corresponding pixel-space representations. Following this operation one may
+        liftn this collection of spin signals to a signal on SO(3) by exploiting the
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:97]: Remove this sentence (see comment above).

File: s2fft/precompute_transforms/custom_ops.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+        L (int): Harmonic band-limit.
+        N (int): Azimuthal band-limit.
+        sampling (str, optional): Sampling scheme.  Supported sampling schemes include
+            {"mw", "mwss"}. Defaults to "mw".
+
+    Returns:
+        np.ndarray: Collection of spin spherical harmonic coefficients with shape :math:`[batch, n_s, L, 2L-1, channels]`.
+
+    """
+    # COMPUTE FFT OVER GAMMA
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:189]: Check spins below azimuthal bandlimit N.

File: s2fft/precompute_transforms/custom_ops.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    x = np.fft.fft(f, axis=-4, norm="forward")
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:190]: Shouldn't we do a DFT here rather than FFT so we only compute the spins of interest?  Rather than computing all and selecting a subset.

Comment (jasonmcewen) [reply s2fft/precompute_transforms/custom_ops.py:190]: Similiarly for the jax implementation below?

File: s2fft/precompute_transforms/custom_ops.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+    L: int,
+    sampling: str = "mw",
+) -> np.ndarray:
+    r"""
+    Transforms from a collection of arbitrary spin signals on the sphere to the
+    corresponding collection of their harmonic coefficients.
+
+        This function takes a multimodal collection of spin spherical harmonic signals
+        on the sphere and transforms them into their spin spherical harmonic coefficients.
+        These cofficients may then be combined into a subset of Wigner coefficients for
```
Comment (jasonmcewen) [inline s2fft/precompute_transforms/custom_ops.py:255]: Typo: coefficients (check for elsewhere as well)

**Review by jasonmcewen — APPROVED** [review]
Looking good @CosmoMatt, just a few minor comments.  Have approved so you can merge once these addressed.

File: s2fft/precompute_transforms/custom_ops.py (hunk 7 above)
Comment (CosmoMatt) [reply s2fft/precompute_transforms/custom_ops.py:189]: Conditional of a traced object may cause loss of performance in this instance. Given this is custom functionality that we are building upon elsewhere, I would recommend we add this catch at a higher level where it won't do any damage (i.e. at the NNX layer definition level).

File: s2fft/precompute_transforms/custom_ops.py (hunk 8 above)
Comment (CosmoMatt) [reply s2fft/precompute_transforms/custom_ops.py:190]: Yep, good catch. I've switched this out to an explicitly DFT.
